python/ray/spark/__init__.py

```diff
@@ -0,0 +1,530 @@
+
+    with open(
+        os.path.join(ray_log_dir, "ray-start-head.log"),
+        "w", buffering=1
+    ) as head_log_fp:
+        ray_head_proc = exec_cmd(
+            ray_head_node_cmd,
+            synchronous=False,
+            capture_output=False,
+            stream_output=False,
+            stdout=head_log_fp,
+            stderr=subprocess.STDOUT,
+        )
+
+    # wait ray head node spin up.
+    wait_ray_node_available(
+        ray_head_hostname, ray_head_port, 40,
+        error_on_failure="Start Ray head node failed!"
+    )
+
+    _logger.info("Ray head node started.")
+
+    # NB:
+    # In order to start ray worker nodes on spark cluster worker machines,
+    # We launch a background spark job:
+    #  1. it is a barrier mode spark job, i.e. all spark tasks in the job runs concurrently.
+    #     if the spark cluster resources are not sufficient to launch all these tasks concurrently,
+    #     the spark job will hang and retry, if exceeding maximum retries, it will fail.
+    #  2. Each spark task launches one ray worker node. This design ensures all ray worker nodes
+    #     has the same shape (same cpus / gpus / memory configuration). If ray worker nodes have
+    #     different shape, the Ray cluster setup will be nondeterministic, and you could get very
+    #     strange results with bad luck on the node sizing.
+    #  3. It starts ray worker node via `ray start` CLI. In each spark task, it creates a
+    #     child process and run `ray start ...` command in blocking mode.
+    #  4. When shut down ray cluster, killing these ray worker nodes is implemented by:
+    #     First, it installs a PR_SET_PDEATHSIG signal for the `ray start ...` child processes
+    #     so that when parent process (pyspark task) dead, the child processes
+    #     (`ray start ...` processes) will receive SIGTERM signal.
+    #     When we need to shut down the ray cluster, call `sparkContext.cancelJobGroup`
+    #     to cancel the background spark job, and it sends SIGKILL signal to all spark tasks,
+    #     so this make spark task processes dead and triggers sending SIGTERM to `ray start ...`
+    #     child processes.
+
+    def ray_cluster_job_mapper(_):
+        from pyspark.taskcontext import BarrierTaskContext
+        _worker_logger = logging.getLogger("ray.spark.worker")
+
+        context = BarrierTaskContext.get()
+        task_id = context.partitionId()
+
+        task_ip_list = [info.address.split(":")[0] for info in context.getTaskInfos()]
+        task_local_rank = get_spark_task_local_rank(task_id, task_ip_list)
+
+        # NB: If we launch multiple ray worker node at the same time,
+        #  it might cause Raylet uses conflicted port.
+        #  probably race conditions issues in ray implementation.
+        #  as a workaround, I add a sleep here to make different local tasks runs starting from
+        #  different time.
+        time.sleep(task_local_rank * 2.0)
```

Comment:
@dbczumar
The workaround alleviates the issue `raylet process unexpectedly exit when multiple ray workers launch on the same node` I talked today.
But if launching too many ray nodes (>=6) on the same spark worker machine, the issue might still happen.
It should be some race condition issues in ray processes.